File: src/main/java/cspfj/constraint/ResidueManagerFast.java
package cspfj.constraint;

import java.util.Arrays;

import cspfj.problem.Variable;

public final class ResidueManagerFast implements ResidueManager {

    private final int[][][] last;

    public ResidueManagerFast(final Variable[] scope) {
        last = new int[scope.length][][];
        for (int i = scope.length; --i >= 0;) {
            last[i] = new int[scope[i].domain().lastIndex() + 1][];
        }
    }

    /*
     * (non-Javadoc)
     * 
     * @see cspfj.constraint.ResidueManager#getResidue(int, int)
     */
    @Override
    public int[] getResidue(final int position, final int index) {
        return last[position][index];
    }

    /*
     * (non-Javadoc)
     * 
     * @see cspfj.constraint.ResidueManager#updateResidue(int[])
     */
    @Override
    public void updateResidue(final int[] residue) {
        for (int position = last.length; --position >= 0;) {
            last[position][residue[position]] = residue;
        }
    }

    @Override
    public void remove(final int[] residue) {
        for (int position = last.length; --position >= 0;) {
            if (Arrays.equals(residue, last[position][residue[position]])) {
                last[position][residue[position]] = null;
            }
        }
    }
}

File: src/main/scala/cspfj/constraint/ResidueManagerFast.scala
package cspfj.constraint;

import scala.annotation.tailrec
import cspfj.problem.Variable

final class ResidueManagerFast(scope: Array[Variable]) {

  val arity = scope.size

  val last = scope map (v => new Array[Array[Int]](v.dom.last + 1))

  def getResidue(position: Int, index: Int) = last(position)(index)

  def updateResidue(residue: Array[Int], i: Int = arity - 1) {
    if (i >= 0) {
      last(i)(residue(i)) = residue
      updateResidue(residue, i - 1)
    }
  }

  def remove(residue: Array[Int], i: Int = arity - 1) {

    if (i >= 0) {
      last(i)(residue(i)) = null
      remove(residue, i - 1)
    }

  }
}
